handlers_test.go

```diff
@@ -0,0 +1,255 @@
+package main
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	btcec "github.com/btcsuite/btcd/btcec/v2"
+	"github.com/stretchr/testify/require"
+	ecrpc "github.com/ziggie1984/Distributed-Mission-Control-for-LND/ecrpc"
+	"go.etcd.io/bbolt"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// generateTestKeys generates a pair of test keys for nodeFrom and nodeTo
+// identity sec compressed pub keys.
+func generateTestKeys(t *testing.T) (nodeFrom, nodeTo []byte) {
+	t.Helper()
+
+	// Generate a private key for nodeFrom.
+	privKeyFrom, err := btcec.NewPrivateKey()
+	if err != nil {
+		t.Fatalf("Failed to generate private key: %v", err)
+	}
+
+	// Serialize the public key of nodeFrom.
+	nodeFrom = privKeyFrom.PubKey().SerializeCompressed()
+
+	// Generate a private key for nodeTo.
+	privKeyTo, err := btcec.NewPrivateKey()
+	if err != nil {
+		t.Fatalf("Failed to generate private key: %v", err)
+	}
+
+	// Serialize the public key of nodeTo.
+	nodeTo = privKeyTo.PubKey().SerializeCompressed()
+
+	// Return the serialized public keys.
+	return nodeFrom, nodeTo
+}
+
+// clearDatabase clears all key-value pairs from the specified bucket.
+func clearDatabase(db *bbolt.DB) error {
+	return db.Update(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(DatabaseBucketName))
+		if b == nil {
+			// If the bucket does not exist, there is nothing to
+			// clear.
+			return nil
+		}
+
+		// Delete all key-value pairs in the bucket.
+		return b.ForEach(func(k, v []byte) error {
+			return b.Delete(k)
+		})
+	})
+}
+
+// TestExternalCoordinatorServer tests the RegisterMissionControl and
+// QueryAggregatedMissionControl methods of the ExternalCoordinatorServer.
+func TestExternalCoordinatorServer(t *testing.T) {
+	tempDir := t.TempDir()
+
+	config := &Config{
+		Database: DatabaseConfig{
+			DatabaseDirPath: tempDir,
+			DatabaseFile:    "test.db",
+			FileLockTimeout: 10 * time.Second,
+			MaxBatchDelay:   10 * time.Millisecond,
+			MaxBatchSize:    1000,
+		},
+	}
+
+	db, err := setupDatabase(config)
+	if err != nil {
+		t.Fatalf("Failed to set up database: %v", err)
+	}
+	defer cleanupDB(db)
+
+	server := NewExternalCoordinatorServer(db)
+
+	t.Run("RegisterMissionControl", func(t *testing.T) {
+		// Case 1: Valid request with nodeFrom, nodeTo and history data.
+		t.Run("ValidRequest", func(t *testing.T) {
```

Comment:
They could start in parallel since we are opening the database once before any test case run and close it after all test cases finish:
https://github.com/mohamedawnallah/Distributed-Mission-Control-for-LND/blob/2a6eea81db8badc02df5535b7999cf34d04e4a7d/handlers_test.go#L75-L79

For write operations could happen in parallel because of batching and read operations there is no lock similar to what exactly happens when we did stress testing.

Reply:
But the point is if we run them in parallel it would lead to inconsistent states between test cases and make them flaky also there is (almost) no saving time when we run those test-suite in parallel:

Test cases running sequentially (current behavior):
```
--- PASS: TestExternalCoordinatorServer (1.10s)
    --- PASS: TestExternalCoordinatorServer/RegisterMissionControl (0.04s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/ValidRequest (0.04s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/NilRequest (0.00s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/EmptyPairs (0.00s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/InvalidNodeFromLength (0.00s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/InvalidNodeToLength (0.00s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/NilHistory (0.00s)
    --- PASS: TestExternalCoordinatorServer/QueryAggregatedMissionControl (1.04s)
        --- PASS: TestExternalCoordinatorServer/QueryAggregatedMissionControl/ValidRequestWithData (1.03s)
        --- PASS: TestExternalCoordinatorServer/QueryAggregatedMissionControl/ValidRequestWithoutData (0.01s)
```


Test cases running in parallel:
```
--- PASS: TestExternalCoordinatorServer (1.10s)
    --- PASS: TestExternalCoordinatorServer/RegisterMissionControl (0.00s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/NilRequest (0.00s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/EmptyPairs (0.00s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/NilHistory (0.03s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/InvalidNodeFromLength (0.03s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/InvalidNodeToLength (0.03s)
        --- PASS: TestExternalCoordinatorServer/RegisterMissionControl/ValidRequest (0.05s)
    --- PASS: TestExternalCoordinatorServer/QueryAggregatedMissionControl (0.00s)
        --- PASS: TestExternalCoordinatorServer/QueryAggregatedMissionControl/ValidRequestWithoutData (0.02s)
        --- PASS: TestExternalCoordinatorServer/QueryAggregatedMissionControl/ValidRequestWithData (1.03s)
```

Reply:
ok cool let's keep them sequential then